src/box/lua/net_box.lua

```diff
@@ -882,6 +921,140 @@ function remote_methods:new_stream()
     return stream
 end
 
+local watcher_methods = {}
+local watcher_mt = {
+    __index = watcher_methods,
+    __newindex = function()
+        box.error(E_PROC_LUA, 'net.box.watcher is ready only')
+    end,
+    __tostring = function()
+        return 'net.box.watcher'
+    end,
+}
+watcher_mt.__serialize = watcher_mt.__tostring
+
+function watcher_methods:_run()
+    local state = self._state
+    assert(state ~= nil)
+    assert(state.has_value)
+    self._version = state.version
+    local status, err = pcall(self._func, state.key, state.value)
+    if not status then
+        log.error(err)
+    end
+    if not self._state then
+        -- The watcher was unregistered while the callback was running.
+        return
+    end
+    assert(state == self._state)
+    if self._version == state.version then
+        -- The value was not updated while this watcher was running.
+        -- Append it to the list of ready watchers and send an ack to
+        -- the server unless we've already sent it.
+        table.insert(state.ready, self)
```

Comment:
> A user application is likely to break somewhere in case of oom, e.g. because it fails to create a string on Lua stack.

It will raise an exception. Like most of our box methods do. User applications somehow deal with that. You can talk to solutions if you want to find more. AFAIK, they try to never fail even in case of OOM to avoid bad errors returned to users or not returning anything. On the top level there always should be a `pcall` which can convert the error into something more user-friendly than a crash and remote connections drop.

> I bet most of our Lua code isn't ready for oom. That's why I think that the program will become unusable.

It will be usable. GC eventually will free more space. Especially if the offender's stack is unrolled and it contained a lot of garbage.

> AFAIU with GC64 Lua should never raise an oom error.

AFAIK, it still is not enabled everywhere, is it?

> With GC32, I think we should terminate the program in case of Lua failing to allocate memory.

I am not against that. But then it must be done before this patch, because otherwise there is a bug.

Reply:
Replaced the table with a singly-linked list. Hope it addresses your concern.

Reply:
It does not fix the entire problem, because, for instance, `self._conn._transport.watch(state.key)` can also raise an exception. Because it calls `netbox_watch_or_unwatch` which uses `luamp_error`. But I rethought my approach and decided that I should not care that much. Besides, I don't know how to fix this in a good way, so I can't recommend anything except C.